### PR TITLE
Add external DNS endpoint CRD

### PR DESCRIPTION
**Pull request by jjngx**
### Proposed changes

This PR introduces ExternalDNS CRD including examples and tests.

### Checklist
Before creating a PR, run through this checklist and mark each as complete.

- [X] I have read the [CONTRIBUTING](https://github.com/nginxinc/kubernetes-ingress/blob/main/CONTRIBUTING.md) doc
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have checked that all unit tests pass after adding my changes
- [X] I have updated necessary documentation (README in examples)
- [X] I have rebased my branch onto main
- [X] I will ensure my PR is targeting the main branch and pulling from my branch from my own fork

### Testing
```
➜  kubernetes-ingress git:(dnsendpoint-crd) make test
go test -tags=aws -shuffle=on -race ./...
ok  	github.com/nginxinc/kubernetes-ingress/cmd/nginx-ingress	1.867s
ok  	github.com/nginxinc/kubernetes-ingress/internal/certmanager	4.629s
?   	github.com/nginxinc/kubernetes-ingress/internal/certmanager/test_files	[no test files]
ok  	github.com/nginxinc/kubernetes-ingress/internal/configs	0.927s
ok  	github.com/nginxinc/kubernetes-ingress/internal/configs/version1	0.575s
ok  	github.com/nginxinc/kubernetes-ingress/internal/configs/version2	0.509s
ok  	github.com/nginxinc/kubernetes-ingress/internal/k8s	0.975s
ok  	github.com/nginxinc/kubernetes-ingress/internal/k8s/appprotect	0.410s
ok  	github.com/nginxinc/kubernetes-ingress/internal/k8s/appprotectcommon	0.320s
ok  	github.com/nginxinc/kubernetes-ingress/internal/k8s/appprotectdos	0.582s
ok  	github.com/nginxinc/kubernetes-ingress/internal/k8s/secrets	0.545s
?   	github.com/nginxinc/kubernetes-ingress/internal/metrics	[no test files]
ok  	github.com/nginxinc/kubernetes-ingress/internal/metrics/collectors	0.873s
ok  	github.com/nginxinc/kubernetes-ingress/internal/nginx	0.919s
?   	github.com/nginxinc/kubernetes-ingress/pkg/apis/configuration	[no test files]
?   	github.com/nginxinc/kubernetes-ingress/pkg/apis/configuration/v1	[no test files]
?   	github.com/nginxinc/kubernetes-ingress/pkg/apis/configuration/v1alpha1	[no test files]
ok  	github.com/nginxinc/kubernetes-ingress/pkg/apis/configuration/validation	0.741s
?   	github.com/nginxinc/kubernetes-ingress/pkg/apis/dos	[no test files]
?   	github.com/nginxinc/kubernetes-ingress/pkg/apis/dos/v1beta1	[no test files]
ok  	github.com/nginxinc/kubernetes-ingress/pkg/apis/dos/validation	0.646s
?   	github.com/nginxinc/kubernetes-ingress/pkg/apis/externaldns	[no test files]
?   	github.com/nginxinc/kubernetes-ingress/pkg/apis/externaldns/v1	[no test files]
ok  	github.com/nginxinc/kubernetes-ingress/pkg/apis/externaldns/validation	0.596s
?   	github.com/nginxinc/kubernetes-ingress/pkg/client/clientset/versioned	[no test files]
?   	github.com/nginxinc/kubernetes-ingress/pkg/client/clientset/versioned/fake	[no test files]
?   	github.com/nginxinc/kubernetes-ingress/pkg/client/clientset/versioned/scheme	[no test files]
?   	github.com/nginxinc/kubernetes-ingress/pkg/client/clientset/versioned/typed/configuration/v1	[no test files]
?   	github.com/nginxinc/kubernetes-ingress/pkg/client/clientset/versioned/typed/configuration/v1/fake	[no test files]
?   	github.com/nginxinc/kubernetes-ingress/pkg/client/clientset/versioned/typed/configuration/v1alpha1	[no test files]
?   	github.com/nginxinc/kubernetes-ingress/pkg/client/clientset/versioned/typed/configuration/v1alpha1/fake	[no test files]
?   	github.com/nginxinc/kubernetes-ingress/pkg/client/clientset/versioned/typed/dos/v1beta1	[no test files]
?   	github.com/nginxinc/kubernetes-ingress/pkg/client/clientset/versioned/typed/dos/v1beta1/fake	[no test files]
?   	github.com/nginxinc/kubernetes-ingress/pkg/client/informers/externalversions	[no test files]
?   	github.com/nginxinc/kubernetes-ingress/pkg/client/informers/externalversions/configuration	[no test files]
?   	github.com/nginxinc/kubernetes-ingress/pkg/client/informers/externalversions/configuration/v1	[no test files]
?   	github.com/nginxinc/kubernetes-ingress/pkg/client/informers/externalversions/configuration/v1alpha1	[no test files]
?   	github.com/nginxinc/kubernetes-ingress/pkg/client/informers/externalversions/dos	[no test files]
?   	github.com/nginxinc/kubernetes-ingress/pkg/client/informers/externalversions/dos/v1beta1	[no test files]
?   	github.com/nginxinc/kubernetes-ingress/pkg/client/informers/externalversions/internalinterfaces	[no test files]
?   	github.com/nginxinc/kubernetes-ingress/pkg/client/listers/configuration/v1	[no test files]
?   	github.com/nginxinc/kubernetes-ingress/pkg/client/listers/configuration/v1alpha1	[no test files]
?   	github.com/nginxinc/kubernetes-ingress/pkg/client/listers/dos/v1beta1	[no test files]
?   	github.com/nginxinc/kubernetes-ingress/tests/test-servers/tcp	[no test files]
?   	github.com/nginxinc/kubernetes-ingress/tests/test-servers/udp	[no test files]
```

With [gotestdox](https://github.com/bitfield/gotestdox) targeting new package:
```
➜  kubernetes-ingress git:(dnsendpoint-crd) gotestdox pkg/apis/externaldns/validation/*.go
command-line-arguments:
 ✔ Validate DNS endpoint with single valid endpoint (0.00s)
 ✔ Validate DNS endpoint with multiple valid endpoints (0.00s)
 ✔ Validate DNS endpoint with multiple valid endpoints and multiple targets (0.00s)
 ✔ Validate DNS endpoint (0.00s)
 ✔ ValidateDNSEndpoint returns error on not supported DNS record type (0.00s)
 ✔ ValidateDNSEndpoint returns error on bogus target hostname (0.00s)
 ✔ ValidateDNSEndpoint returns error on duplicated target (0.00s)
 ✔ ValidateDNSEndpoint returns error on bogus ttl record (0.00s)
 ✔ ValidateDNSEndpoint returns error on bogus dns name (0.00s)
 ✔ ValidateDNSEndpoint returns error on empty dns name (0.00s)
 ✔ ValidateDNSEndpoint returns error on bogus target name (0.00s)
 ✔ ValidateDNSEndpoint returns error on empty target name (0.00s)
 ✔ ValidateDNSEndpoint returns error on bogus target name# 01 (0.00s)
 ✔ ValidateDNSEndpoint returns error on empty slice of endpoints (0.00s)
 ✔ ValidateDNSEndpoint returns error on (0.00s)
```

With local ```kind``` cluster:
```
➜  kubernetes-ingress git:(dnsendpoint-crd) kubectl create -f deployments/common/crds/externaldns.nginx.org_dnsendpoints.yaml
customresourcedefinition.apiextensions.k8s.io/dnsendpoints.externaldns.nginx.org created
```

```
➜  kubernetes-ingress git:(externaldns) ✗ kubectl create -f dnstest.yaml
dnsendpoint.externaldns.nginx.org/dnsundertest created
```
where ```dnstest.yaml```:
```yaml
apiVersion: externaldns.nginx.org/v1
kind: DNSEndpoint
metadata:
  name: dnsundertest
spec:
  endpoints:
  - dnsName: example.co.uk
    recordTTL: 3600
    recordType: A
    targets:
    - 192.168.99.216
  - dnsName: example.ie
    recordTTL: 1800
    recordType: A
    targets:
    - 192.168.2.3
```

```
➜  kubernetes-ingress git:(dnsendpoint-crd) kubectl get dnsendpoints
NAME           AGE
dnsundertest   4m25s
```
```
➜  kubernetes-ingress git:(dnsendpoint-crd) kubectl describe dnsendpoints dnsundertest
Name:         dnsundertest
Namespace:    default
Labels:       <none>
Annotations:  <none>
API Version:  externaldns.nginx.org/v1
Kind:         DNSEndpoint
Metadata:
  Creation Timestamp:  2022-06-16T12:45:15Z
  Generation:          1
  Managed Fields:
    API Version:  externaldns.nginx.org/v1
    Fields Type:  FieldsV1
    fieldsV1:
      f:spec:
        .:
        f:endpoints:
    Manager:         kubectl-create
    Operation:       Update
    Time:            2022-06-16T12:45:15Z
  Resource Version:  1548492
  UID:               50330dea-0108-4a67-892d-cc46c5e52d1c
Spec:
  Endpoints:
    Dns Name:     example.co.uk
    Record TTL:   3600
    Record Type:  A
    Targets:
      192.168.99.216
    Dns Name:     example.ie
    Record TTL:   1800
    Record Type:  A
    Targets:
      192.168.2.3
Events:  <none>
```

### CODE REVIEW

File: examples/custom-resources/external-dns/README.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+Step 1: Register the external-crd with the k8s api (run from the root of this repo):
+
+    ```k apply -f deployments/common/crd/externaldns.nginx.org_dnsendpoints.yaml```
```
Comment (ciarams87) [inline examples/custom-resources/external-dns/README.md:3]:
```suggestion
    ```k apply -f deployments/common/crds/externaldns.nginx.org_dnsendpoints.yaml```
```
```

**Review by ciarams87 — CHANGES_REQUESTED** [review]
`make update-codegen` needs to be ran to generate the client code

Edit:

The `hack/update-codegen` file needs to be updated on line 36 to generate this code from 
`  "configuration:v1alpha1,v1 dos:v1beta1" \` to
`  "configuration:v1alpha1,v1 dos:v1beta1 externaldns:v1" \ `

File: pkg/apis/externaldns/validation/externaldns.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+	}
+	return nil
+}
+
+var (
+	// validRecords represents allowed DNS record names
+	//
+	// NGINX ingress controller at the moment supports
+	// a subset of DNS record types listed in the external-dns project.
+	validRecords = []string{"A", "CNAME"}
```
Comment (ciarams87) [inline pkg/apis/externaldns/validation/externaldns.go:109]: Should we also support AAAA names?

Comment (jjngx) [reply pkg/apis/externaldns/validation/externaldns.go]: Allowed ```AAAA``` records and added IPv6 target validation

**Pull request merged**